### PR TITLE
DB sources data type support - adding new data types 

### PR DESCRIPTION
**Pull request by akashkulk**
Breaking up https://github.com/airbytehq/airbyte/pull/19403 into smaller pieces. Specifically, this PR just refactors code to support the new data types:

- `JsonSchemaPrimitiveUtil.java` & `JsonSchemaType.java` : New data types associated with the new data types WellKnownTypes are added here. When most of the code has migrated to the new data types, we can remove the old data types. Also integrating `JsonSchemaReferenceTypes` from Edward's [PR](https://github.com/airbytehq/airbyte/pull/19240), so will have to resolve merge conflicts at some point. 
- `IncrementalUtils.java` : Support for new data types
- Everything else : Refactoring

### CODE REVIEW

File: airbyte-protocol/protocol-models/src/main/java/io/airbyte/protocol/models/JsonSchemaPrimitiveUtil.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+import com.google.common.collect.BiMap;
+import com.google.common.collect.ImmutableBiMap;
+import com.google.common.collect.ImmutableSet;
+import java.util.Set;
+
+public class JsonSchemaPrimitiveUtil {
+
+  public enum JsonSchemaPrimitive {
+
```
Comment (rodireich) [inline airbyte-protocol/protocol-models/src/main/java/io/airbyte/protocol/models/JsonSchemaPrimitiveUtil.java:26]: nit: mention this is V0

File: airbyte-protocol/protocol-models/src/main/java/io/airbyte/protocol/models/JsonSchemaType.java [hunk 2]
```diff
@@ -72,7 +88,11 @@ public static class Builder {
 
     private Builder(final JsonSchemaPrimitive type) {
       typeMapBuilder = ImmutableMap.builder();
-      typeMapBuilder.put(TYPE, type.name().toLowerCase());
+      if (JsonSchemaPrimitiveUtil.isV0Schema(type)) {
+        typeMapBuilder.put(TYPE, type.name().toLowerCase());
+      } else {
+        typeMapBuilder.put(REF, PRIMITIVE_TO_REFERENCE_BIMAP.get(type));
```
Comment (rodireich) [inline airbyte-protocol/protocol-models/src/main/java/io/airbyte/protocol/models/JsonSchemaType.java:94]: Should there be a `toLowercase()`?

**Review by edgao — APPROVED** [review]
nice - didn't do a super deep review since I already reviewed https://github.com/airbytehq/airbyte/pull/19403; lmk if I should take a closer look.

I'm still tweaking JsonSchemaReferenceTypes (mostly for implementing the downgrade migration) so there's definitely going to be some conflicts once that's merged in :P but we can deal with that later

File: airbyte-protocol/protocol-models/src/main/java/io/airbyte/protocol/models/JsonSchemaType.java (hunk 2 above)
Comment (akashkulk) [reply airbyte-protocol/protocol-models/src/main/java/io/airbyte/protocol/models/JsonSchemaType.java:94]: No, the string that `PRIMITIVE_TO_REFERENCE_MAP` returns is already lowercase

File: airbyte-protocol/protocol-models/src/main/java/io/airbyte/protocol/models/JsonSchemaPrimitiveUtil.java (hunk 1 above)
Comment (akashkulk) [reply airbyte-protocol/protocol-models/src/main/java/io/airbyte/protocol/models/JsonSchemaPrimitiveUtil.java:26]: Done

**Comment by akashkulk** [comment]
> nice - didn't do a super deep review since I already reviewed #19403; lmk if I should take a closer look.
> 
> I'm still tweaking JsonSchemaReferenceTypes (mostly for implementing the downgrade migration) so there's definitely going to be some conflicts once that's merged in :P but we can deal with that later

Ack - I might be there first, so hopefully the conflicts aren't too bad :)

**Review by rodireich — APPROVED** [review]

**Pull request merged**